docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,401 @@
+  string codespace = 8;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should allow for configurable buffering of a reasonable, yet small,
+number of requests to the companion service. If this buffer fills up, we should
+panic.
+
+This data would need to be stored on disk to ensure that Tendermint could
+attempt to resubmit all unsubmitted data upon Tendermint startup. At present,
+this data is already stored on disk, and so practically we would need to
+implement some form of background pruning mechanism to remove the data we know
+has been shared with the companion service.
+
+### Interaction with Block Sync
+
+During block sync, request buffering should be disabled and requests to the
+companion service should be blocking.
+
+### Configuration
+
+The following configuration file update is proposed to support the data
+companion API.
+
+```toml
+[data_companion]
+# By default, the data companion service interaction is disabled. It is
+# recommended that this only be enabled on full nodes and not validators so as
+# to minimize the possibility of network instability.
+enabled = false
+
+# Address at which the gRPC companion service server is hosted. It is
+# recommended that this companion service be co-located at least within the same
+# data center as the Tendermint node to reduce the risk of network latencies
+# interfering in node operation.
+addr = "http://localhost:26659"
+
+# The number of requests to the companion service to durably buffer. Set to 0 to
+# enable blocking send mode, which will block on every request to the companion
+# service. A sensible non-zero value would be 10 (if the companion service is
+# unavailable to receive 10 blocks' data, something is probably wrong and
+# requires intervention).
+buffer_size = 0
+
+# Use the experimental gzip compression call option when submitting data to the
+# server. See https://pkg.go.dev/google.golang.org/grpc#UseCompressor
+experimental_use_gzip = false
+```
+
+It is unclear at present whether compressing the data being sent to the
+companion service will result in meaningful benefits.
+
+### Monitoring
+
+To monitor the health of the interaction between Tendermint and the companion
+service, the following additional Prometheus metrics are proposed:
+
+- `data_companion_send_time` - A gauge that indicates the maximum time, in
+  milliseconds, taken to send a single request to the companion service take
+  from a rolling window of tracked send times (e.g. the maximum send time over
+  the past minute).
+- `data_companion_buffer_utilization` - A gauge indicating how much of the
+  request buffer has been used.
+
+## Implications
+
+1. We will be able to mark the following RPC APIs as deprecated and schedule
+   them for removal in a future release:
+   1. The [WebSocket subscription API][websocket-api]
+   2. [`/tx_search`]
+   3. [`/block_search`]
+   4. [`/broadcast_tx_commit`]
```

Comment:
I think this endpoint requires more discussion. I think there's utility in having a tiny index (perhaps builtin to to the blockstore) that is able to let a user know that their transaction has been committed. Perhaps we still deprecate this (as we'd discussed previously) but I think we should still support the use case of I submit a transaction and I poll to check that it is actually committed. Perhaps this requires a broader discussion on transaction tracing